File: packages/dvai-bridge-capacitor-llama/ios/Tests/DVAICapacitorLlamaTests/LlamaCppBridgeTest.swift
import XCTest
@testable import DVAICapacitorLlama
import DVAICapacitorLlamaObjC

final class LlamaCppBridgeTest: XCTestCase {
    func testInitiallyNotLoaded() {
        let bridge = LlamaCppBridge()
        XCTAssertFalse(bridge.isLoaded)
        XCTAssertNil(bridge.currentModelPath)
    }

    func testLoadEmptyPathFails() {
        let bridge = LlamaCppBridge()
        XCTAssertThrowsError(
            try bridge.loadModel(
                atPath: "",
                mmprojPath: nil,
                gpuLayers: 99,
                contextSize: 2048,
                threads: 4,
                embeddingMode: false
            )
        )
        XCTAssertFalse(bridge.isLoaded)
    }

    func testLoadFakePathFails() {
        let bridge = LlamaCppBridge()
        XCTAssertThrowsError(
            try bridge.loadModel(
                atPath: "/tmp/definitely-does-not-exist.gguf",
                mmprojPath: nil,
                gpuLayers: 99,
                contextSize: 2048,
                threads: 4,
                embeddingMode: false
            )
        )
        XCTAssertFalse(bridge.isLoaded)
        XCTAssertNil(bridge.currentModelPath)
    }

    func testVersionStringContainsLlama() {
        let bridge = LlamaCppBridge()
        let version = bridge.versionString()
        let prefix = "llama.cpp "
        XCTAssertTrue(
            version.hasPrefix(prefix),
            "expected versionString to start with '\(prefix)', got: \(version)"
        )
        XCTAssertGreaterThan(
            version.count,
            prefix.count,
            "expected versionString to include system info after prefix, got: \(version)"
        )
    }

    // MARK: - Multimodal stubs (Phase 2A Pass 1)

    func testInitiallyMmprojNotLoaded() {
        let bridge = LlamaCppBridge()
        XCTAssertFalse(bridge.isMmprojLoaded)
    }

    func testLoadMmprojRequiresMainModel() {
        let bridge = LlamaCppBridge()
        // Main model never loaded -> should fail with code 31.
        XCTAssertThrowsError(try bridge.loadMmproj(atPath: "/tmp/fake-mmproj.gguf"))
        XCTAssertFalse(bridge.isMmprojLoaded)
    }

    func testEmptyMmprojPathFails() {
        let bridge = LlamaCppBridge()
        XCTAssertThrowsError(try bridge.loadMmproj(atPath: ""))
        XCTAssertFalse(bridge.isMmprojLoaded)
    }

    func testUnloadMmprojIsIdempotent() {
        let bridge = LlamaCppBridge()
        // Repeated unload calls on a never-loaded bridge must not crash.
        bridge.unloadMmproj()
        bridge.unloadMmproj()
        XCTAssertFalse(bridge.isMmprojLoaded)
    }

    // MARK: - Multimodal — Phase 2A Pass 2

    /// Without a loaded model, completeMultimodalPrompt must return an error
    /// (code 50, "Model not loaded"). We can verify this without a real model.
    func testCompleteMultimodalRequiresModelLoaded() {
        let bridge = LlamaCppBridge()
        XCTAssertFalse(bridge.isLoaded)
        XCTAssertThrowsError(
            try bridge.completeMultimodalPrompt(
                "ignored <__media__>",
                media: [Data([0x00, 0x01, 0x02])],
                maxTokens: 8,
                temperature: 0.0,
                topP: 1.0
            )
        ) { error in
            let nsErr = error as NSError
            XCTAssertEqual(nsErr.domain, "DVAIBridgeLlama")
            XCTAssertEqual(nsErr.code, 50, "expected code 50 (Model not loaded)")
        }
    }

    /// applyChatTemplate without a loaded model returns the "Model not loaded"
    /// error (code 40). Same can-run-without-model property as above.
    func testApplyChatTemplateRequiresModelLoaded() {
        let bridge = LlamaCppBridge()
        XCTAssertFalse(bridge.isLoaded)
        XCTAssertThrowsError(
            try bridge.applyChatTemplate(
                nil,
                messages: [["role": "user", "content": "hi"]],
                addAssistant: true
            )
        ) { error in
            let nsErr = error as NSError
            XCTAssertEqual(nsErr.domain, "DVAIBridgeLlama")
            XCTAssertEqual(nsErr.code, 40, "expected code 40 (Model not loaded)")
        }
    }

    /// hasAudioEncoder is always false when no mmproj is loaded.
    func testHasAudioEncoderFalseWithoutMmproj() {
        let bridge = LlamaCppBridge()
        XCTAssertFalse(bridge.hasAudioEncoder())
    }
}

File: packages/dvai-bridge-android-llama-core/android/src/test/java/co/deepvoiceai/bridge/llama/core/LlamaCppBridgeTest.kt
package co.deepvoiceai.bridge.llama.core

import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

/**
 * JVM unit tests for LlamaCppBridge's pure-Kotlin state machine.
 * JNI methods aren't loadable here (no .so on JVM classpath); that's
 * exercised by instrumented tests on a device/emulator (later tier).
 *
 * Robolectric is used so System.loadLibrary's UnsatisfiedLinkError
 * is caught gracefully (the bridge falls through to a stub state).
 */
@RunWith(RobolectricTestRunner::class)
class LlamaCppBridgeTest {

    @Test
    fun `initially not loaded`() {
        val bridge = LlamaCppBridge()
        assertFalse(bridge.isLoaded())
        assertNull(bridge.getCurrentModelPath())
    }

    @Test
    fun `loadModel with empty path returns false`() {
        val bridge = LlamaCppBridge()
        val ok = bridge.loadModel(
            path = "",
            mmprojPath = null,
            gpuLayers = 99,
            contextSize = 2048,
            threads = 4,
            embeddingMode = false,
        )
        assertFalse(ok)
        assertFalse(bridge.isLoaded())
    }

    @Test
    fun `loadModel + unload cycle updates state`() {
        val bridge = LlamaCppBridge()
        val ok = bridge.loadModel(
            path = "/tmp/fake.gguf",
            mmprojPath = null,
            gpuLayers = 99,
            contextSize = 2048,
            threads = 4,
            embeddingMode = false,
        )
        assertTrue(ok)
        assertTrue(bridge.isLoaded())
        assertEquals("/tmp/fake.gguf", bridge.getCurrentModelPath())

        bridge.unload()
        assertFalse(bridge.isLoaded())
        assertNull(bridge.getCurrentModelPath())
    }

    @Test
    fun `versionString returns stub identifier when JNI unavailable`() {
        val bridge = LlamaCppBridge()
        // In JVM tests, JNI isn't loadable, so the Kotlin fallback is returned.
        assertEquals("llama.cpp-stub-android-0.1", bridge.versionString())
    }

    @Test
    fun `completePrompt returns null when not loaded`() {
        val bridge = LlamaCppBridge()
        // No loadModel call -- bridge stays unloaded.
        val result = bridge.completePrompt(
            prompt = "hello",
            maxTokens = 8,
            temperature = 0.7f,
            topP = 0.9f,
        )
        assertNull(result)
    }

    @Test
    fun `completePrompt returns null on JVM fallback even after loadModel`() {
        val bridge = LlamaCppBridge()
        val ok = bridge.loadModel(
            path = "/tmp/fake.gguf",
            mmprojPath = null,
            gpuLayers = 99,
            contextSize = 2048,
            threads = 4,
            embeddingMode = false,
        )
        // On JVM (no .so), the existing fallback returns ok=true.
        assertTrue(ok)
        assertTrue(bridge.isLoaded())
        // But completePrompt's UnsatisfiedLinkError catch returns null.
        val result = bridge.completePrompt("hi", 4, 0.0f, 1.0f)
        assertNull(result)
    }

    // ---- Multimodal stubs (Phase 2A Pass 1) ----

    @Test
    fun `loadMmproj fails when main model not loaded`() {
        val bridge = LlamaCppBridge()
        // No loadModel call -- bridge stays unloaded.
        val result = bridge.loadMmproj("/tmp/fake-mmproj.gguf")
        assertFalse(result)
        assertFalse(bridge.isMmprojLoaded())
    }

    @Test
    fun `loadMmproj with empty path returns false`() {
        val bridge = LlamaCppBridge()
        bridge.loadModel(
            path = "/tmp/fake.gguf",
            mmprojPath = null,
            gpuLayers = 99,
            contextSize = 2048,
            threads = 4,
            embeddingMode = false,
        )
        val result = bridge.loadMmproj("")
        assertFalse(result)
        assertFalse(bridge.isMmprojLoaded())
    }

    // ---- Phase 2A Pass 2 ----

    @Test
    fun `hasAudioEncoder is false without mmproj`() {
        val bridge = LlamaCppBridge()
        assertFalse(bridge.hasAudioEncoder())
    }

    @Test
    fun `applyChatTemplate returns null when not loaded`() {
        val bridge = LlamaCppBridge()
        val result = bridge.applyChatTemplate(
            templateOverride = null,
            messages = listOf(mapOf("role" to "user", "content" to "hi")),
            addAssistant = true,
        )
        assertNull(result)
    }

    @Test
    fun `completeMultimodalPrompt returns null when mmproj not loaded`() {
        val bridge = LlamaCppBridge()
        bridge.loadModel(
            path = "/tmp/fake.gguf",
            mmprojPath = null,
            gpuLayers = 99,
            contextSize = 2048,
            threads = 4,
            embeddingMode = false,
        )
        // Loaded but no mmproj -> bridge returns null without invoking JNI.
        val result = bridge.completeMultimodalPrompt(
            prompt = "hi <__media__>",
            media = listOf(byteArrayOf(0x00, 0x01)),
            maxTokens = 8,
            temperature = 0.0f,
            topP = 1.0f,
        )
        assertNull(result)
    }

    @Test
    fun `loadMmproj + unload cycle updates mmproj state`() {
        val bridge = LlamaCppBridge()
        bridge.loadModel(
            path = "/tmp/fake.gguf",
            mmprojPath = null,
            gpuLayers = 99,
            contextSize = 2048,
            threads = 4,
            embeddingMode = false,
        )
        // JVM fallback: returns true.
        val ok = bridge.loadMmproj("/tmp/fake-mmproj.gguf")
        assertTrue(ok)
        assertTrue(bridge.isMmprojLoaded())

        bridge.unloadMmproj()
        assertFalse(bridge.isMmprojLoaded())
        // Idempotent: a second unload should not throw or flip state back.
        bridge.unloadMmproj()
        assertFalse(bridge.isMmprojLoaded())
    }
}
